File: Custom_Yolo/Utils/decode_net_out.py
#Function to transform the output of the network in the correct format
def decode_net_out(out):
  # IN : out [7x7x5]
  # RETURNS : bboxes [list] of the best training boxes
  cell_size = 448/7
  bboxes = []
  for i in range(0,7):
    for j in range(0,7):
      if out[i][j][0] > 0.7: #CONFIDENCE THRESHOLD (P of the cell)
        bbox = np.zeros((4))
        ox_cell = out[i][j][1]
        oy_cell = out[i][j][2]
        w_cell = out[i][j][3]
        h_cell = out[i][j][4]
        if(ox_cell != 0 and oy_cell !=0):
          ox = trunc(ox_cell*cell_size + cell_size*j)
          oy = trunc(oy_cell*cell_size + cell_size*i)
          w = w_cell*cell_size
          h = h_cell*cell_size
          lx = ox - w/2
          ly = oy - h/2
          bbox = [out[i][j][0], int(lx),int(ly),int(w),int(h)]
          bboxes.append(bbox)
  return bboxes

File: Custom_Yolo/decodeNetOut.cpp
#include <opencv2/core.hpp>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/objdetect.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/dnn/dnn.hpp>
#include "classes.h"

using namespace std;
using namespace cv;
using namespace dnn;

void decodeNetOut(cv::Mat output, cv::Mat image)
{
    typedef Vec<float, 5> Vec5f; //definition of template to get the outputs of the network
    
    int cell_size = 64; //Cell size is 448/7, needed to convert the values of the rect
    
    //Iterate over each grid cell to find the bboxes
    for(int i = 0; i < 7; i++)
    {
        for(int j = 0; j < 7; j++)
        {
            if(output.at<Vec5f>(i,j)[0] > 0.6) //Keep only boxes with a confidence bigger than 0.6
            {
                float x = output.at<Vec5f>(i,j)[1];
                float y = output.at<Vec5f>(i,j)[2];
                float w_cell = output.at<Vec5f>(i,j)[3];
                float h_cell = output.at<Vec5f>(i,j)[4];
                
                double ox = x*cell_size + cell_size*j;
                double oy = y*cell_size + cell_size*i;
                double w = w_cell*448;
                double h = h_cell*448;
                double lx = ox - w/2;
                double ly = oy - h/2;
                double bx = ox + w/2;
                double by = oy + h/2;
                rectangle(image, Point(static_cast<int>(lx),static_cast<int>(ly)), Point(static_cast<int>(bx),static_cast<int>(by)), Scalar(0,255,0), 1, LINE_8);
            }
        }
    }

    namedWindow("Image", cv::WINDOW_AUTOSIZE);
    imshow("Image", image);
    waitKey(0);

}
